src/Storages/Hive/StorageHive.cpp

```diff
@@ -0,0 +1,666 @@
+                auto format = FormatFactory::instance().getInputFormat(
+                    "CSV", buffer, partition_key_expr->getSampleBlock(), getContext(), getContext()->getSettingsRef().max_block_size);
+                auto pipeline = QueryPipeline(std::move(format));
+                auto reader = std::make_unique<PullingPipelineExecutor>(pipeline);
+                Block block;
+                if (!reader->pull(block) || !block.rows())
+                    throw Exception("Could not parse partition value: " + wb.str(), ErrorCodes::INVALID_PARTITION_VALUE);
+
+                FieldVector fields(partition_names.size());
+                for (size_t i = 0; i < partition_names.size(); ++i)
+                {
+                    block.getByPosition(i).column->get(0, fields[i]);
+                    ranges.emplace_back(fields[i]);
+                }
+
+                const KeyCondition partition_key_condition(query_info, getContext(), partition_names, partition_minmax_idx_expr);
+                if (!partition_key_condition.checkInHyperrectangle(ranges, partition_types).can_be_true)
+                {
+                    //LOG_DEBUG(log, "skip partition:{}", boost::algorithm::join(p.values, "|"));
+                    return;
+                }
+
+                LOG_TRACE(log, "list location:{}", p.sd.location);
+                auto paths = list_paths(p.sd.location);
+                for (const auto & path : paths)
+                {
+                    append_hive_files(path, fields);
+                }
+            };
+            pool.scheduleOrThrowOnError(f);
+        }
+        pool.wait();
+    }
+    else if (partition_name_types.empty()) // Partition keys is empty
+    {
+        auto paths = list_paths(hive_table_metadata->getTable()->sd.location);
+        for (const auto & path : paths)
+        {
+            pool.scheduleOrThrowOnError([&] { append_hive_files(path, {}); });
+        }
+        pool.wait();
+    }
+    else // Partition keys is not empty but partitions is empty
+    {
+        return {};
+    }
+
+    auto sources_info = std::make_shared<HiveSource::SourcesInfo>();
+    sources_info->hive_files = std::move(hive_files);
+    sources_info->database = hive_database;
+    sources_info->table_name = hive_table;
+    sources_info->hive_metastore_client = hive_metastore_client;
+    sources_info->partition_name_types = partition_name_types;
+    for (const auto & column : column_names)
+    {
+        if (column == "_path")
+            sources_info->need_path_column = true;
+        if (column == "_file")
+            sources_info->need_file_column = true;
+    }
+
+    if (num_streams > sources_info->hive_files.size())
+        num_streams = sources_info->hive_files.size();
+
+    Pipes pipes;
+    for (size_t i = 0; i < num_streams; ++i)
+    {
+        pipes.emplace_back(std::make_shared<HiveSource>(
+            sources_info,
+            hdfs_namenode_url,
```

Comment:
Why not just give every source a needed uri instead of having atomic `next_read_uri`? Looks like their sizes must be equal?

Reply:
I think the load between pipes is more balanced using atomic `next_uri_to_read` than giving files of equal size to pipes. Consider HDFS files have different sizes, whenever any pipe finish reading current file, it can fetch next file and continue.